src/app/data-planes/data/index.ts

```diff
@@ -13,56 +17,138 @@ export type DataplaneWarning = {
-  const subscriptions = dataplaneOverview.dataplaneInsight?.subscriptions ?? []
-  if (subscriptions.length === 0) {
-    return undefined
+export type DataplaneNetworking = PartialDataplaneNetworking
+
+export type Dataplane = PartialDataplane & {
+  networking: DataplaneNetworking
+}
+
+export type DataplaneInsight = PartialDataplaneInsight & {
+  subscriptions: DiscoverySubscription[]
+  connectedSubscription?: DiscoverySubscription
+}
+
+export type DataplaneOverview = PartialDataplaneOverview & {
+  dataplane: {
+    networking: DataplaneNetworking
   }
+  dataplaneInsight: DataplaneInsight
+  dataplaneType: 'standard' | 'builtin' | 'delegated'
+  status: 'online' | 'offline' | 'partially_degraded'
+  unhealthyInbounds: Array<{ service: string, port: number }>
+  lastUpdateTime?: string | undefined
+  warnings: DataplaneWarning[]
+  isCertExpired: boolean
+  serviceTags: LabelValue[]
+}
+
+export const DataplaneNetworking = {
+  fromObject(partialDataplaneNetworking: PartialDataplaneNetworking): DataplaneNetworking {
+    // TODO: Normalize inbound and outbound omitempty arrays.
+    // TODO: Consider to determine inbound address and service address here.
+    // Address: `${inbound.address ?? props.data.dataplane.networking.advertisedAddress ?? props.data.dataplane.networking.address}:${inbound.port}`
+    // Service address: `${inbound.serviceAddress ?? inbound.address ?? props.data.dataplane.networking.address}:${inbound.servicePort ?? inbound.port}`
+    return {
+      ...partialDataplaneNetworking,
+    }
+  },
+}
+
+export const Dataplane = {
+  fromObject(partialDataplane: PartialDataplane): Dataplane {
+    return {
+      ...partialDataplane,
+      networking: DataplaneNetworking.fromObject(partialDataplane.networking),
+    }
+  },
+}
+
+export const DataplaneInsight = {
+  fromObject(partialDataplaneInsight: PartialDataplaneInsight | undefined): DataplaneInsight {
+    const subscriptions = Array.isArray(partialDataplaneInsight?.subscriptions) ? partialDataplaneInsight.subscriptions : []
+    const connectedSubscription = subscriptions.find((subscription) => !subscription.disconnectTime)
+
+    return {
+      ...partialDataplaneInsight,
+      connectedSubscription,
+      subscriptions,
+    }
+  },
+}
 
-  const lastSubscription = subscriptions[subscriptions.length - 1]
-  return lastSubscription.status.lastUpdateTime
+export const DataplaneOverview = {
+  fromObject(partialDataplaneOverview: PartialDataplaneOverview, canUseZones: boolean): DataplaneOverview {
+    const dataplaneInsight = DataplaneInsight.fromObject(partialDataplaneOverview.dataplaneInsight)
+    const networking = DataplaneNetworking.fromObject(partialDataplaneOverview.dataplane.networking)
+
+    const dataplaneType = getDataplaneType(networking)
+    const status = getStatus(partialDataplaneOverview)
+    const unhealthyInbounds = getUnhealthyInbounds(networking)
+    const lastUpdateTime = dataplaneInsight.subscriptions.at(-1)?.status.lastUpdateTime
```

Comment:
The `lastUpdateTime` should also yield a result for disconnected Dataplanes and that’s not possible with the concept of `connectedSubscription`.

Reply:
In the same sense that we shouldn't be looking for the last subscription elsewhere, should we be finding the one/subscription with the latest `lastUpdateTime` here?

If so, I'm not super precious about getting this tweaked in this PR just if we are doing this sort of thing we should do it here or in a follow up.

Let me know on the above question.

Reply:
> In the same sense that we shouldn't be looking for the last subscription elsewhere, should we be finding the one/subscription with the latest `lastUpdateTime` here?

To determine what’s the latest subscription by `lastUpdateTime`, we would have to iterate over all subscriptions, convert the `lastUpdateTime` to `Date` objects and keep record of the latest one.

If we do this here (for `lastUpdateTime`), we should also do it for determining whether we can set a `connectedSubscription`: taking the latest subscription and checking if that subscription has no `disconnectTime`. If so, that’s the `connectedSubscription`.

As far as I understood it, going back to the mode of discovering the connected subscription by finding the _first_ not disconnected one (in contrast to just checking for that condition on the last subscription) was meant as a safe guard in case we encounter bugs of the nature where subscriptions were out of order.

If now we introduce the new measure of only looking at the subscription with the latest `lastUpdateTime`, we potentially wouldn’t find the a not disconnected subscription if it doesn’t happen to be the latest.

To avoid _that_, we would have to perform the searches like this:

- For `lastUpdateTime`: find the subscription with the latest `lastUpdateTime` overall (and take its `lastUpdateTime`)
- For `connectedSubscription`: find the _not disconnected_ subscription with the latest `lastUpdateTime`

Is this what we want? It seems quite convoluted.

Reply:
I think there are some very good points here.

> we would have to iterate over all subscriptions, convert the lastUpdateTime to Date objects 

I would like to convert all our Dates to Date objects anyway.

> If we do this here (for lastUpdateTime), we should also do it for determining whether we can set a connectedSubscription: taking the latest subscription and checking if that subscription has no disconnectTime. If so, that’s the connectedSubscription.

🤔 so you are saying that we can potentially have two subscriptions without a disconnectTime? I think the difference here is the phrase `lastUpdate`. If the one connected subscription can be anywhere in the list I would have guessed the `lastUpdateTime` also could be anywhere in the list?

Lets resolve this one if you are ok doing that, I would like to look/ask/think about this further outside of this PR, but I would guess this will come up again.

Reply:
> Lets resolve this one if you are ok doing that, I would like to look/ask/think about this further outside of this PR, but I would guess this will come up again.

Fine by me.

> 🤔 so you are saying that we can potentially have two subscriptions without a disconnectTime?

At least in the bugs we’ve seen, yes, that was part of the problem: not only were the subscriptions out of order, there were multiple not disconnected subscriptions.

Everything being how it’s supposed to be however would mean (1) there is only ever one not disconnected subscription and (2) the last subscription is the only interesting/relevant one to look at.

Consider also the detail that the API will actually not populate some of the data of older subscriptions (e.g. the config in the case of `ZoneOverview`). I think it would be good to have all of those rules written down somewhere that we can look at (preferably as close to the source (i.e. the backend) as possible).

Reply:
ok gotcha, definitely worth thinking about this one further, I think there will be more work to come around here, not only here but in zones also. Lets resolve this one for now.

Reply:
Just for reference, we could consider adding the following two functions to the subscriptions data layer for this purpose:

```ts
export function getLatestSubscription<S extends KDSSubscription | DiscoverySubscription>(subscriptions: S[]): S | undefined {
  let latestSubscription: S | undefined

  let currentLastUpdateTime = 0
  for (const subscription of subscriptions) {
    const lastUpdateTime = (new Date(subscription.status.lastUpdateTime)).getTime()

    if (lastUpdateTime > currentLastUpdateTime) {
      currentLastUpdateTime = lastUpdateTime
      latestSubscription = subscription
    }
  }

  return latestSubscription
}

export function getConnectedSubscription<S extends KDSSubscription | DiscoverySubscription>(subscriptions: S[]): S | undefined {
  const latestSubscription = getLatestSubscription(subscriptions)

  if (latestSubscription && !latestSubscription.disconnectTime) {
    return latestSubscription
  }

  return undefined
}
```

And store both `latestSubscription` and `connectedSubscription` on a `*Insight` object (both for ZoneOverview and DataplaneOverview). This would isolate both modules from changes when we decide to change the rules.

Reply:
> And store both latestSubscription and connectedSubscription on a *Insight object (both for ZoneOverview and DataplaneOverview). 

Yes exactly that ^, please don't go ahead and do it just yet. I'd like to get everything moved over to this data layer `fromObject` API first. Then we can do these kind of things separately as a whole across the entire project/all entities.